Change public accounts pages to mount the web UI

Gargron opened the pull request:
- Remove `/web` prefix from web app (maintain redirect)
- Redirect any non-local resource to its permalink (accounts, statuses)
- Pre-render a noindex tag on all web app pages except:
  - Profiles
  - Statuses
  - About
  - Privacy Policy
  - Explore (while it's rendered through the `HomeController`, we make an exception when `request.path == '/'`)
- Override the noindex tag in the web app for the same pages when a search engine navigates within the web app
- Ensure that the web app renders a noindex tag on error pages, including error boundary
- Remove remote follow / remote interaction dialogs (superceded by in-app interaction modal)

___

This finally completes #16800

___

Note about `robots` meta tag handling: `react-helmet` does not overwrite header tags that were already in the HTML, it only appends new ones. That means if the HTML sets one `robots` tag, it will be present on any subsequent SPA navigation, even if other routes append their own `robots` tag. I am not 100% sure how such a case is handled by search engines -- which tag is prioritized, the first, or the last. That being said, according to Google's documentation, Google will only render and process the page's JavaScript if the page does not have a `noindex` directive on it. That means we only have to worry about SPA-rendered `robots` tags if the *initial* tag was *not* `noindex`.

Gargron committed:
Change public accounts pages to mount the web UI

ykzts reviewed (changes requested):
```console
git rm spec/controllers/remote_follow_controller_spec.rb
```

ClearlyClaire commented:
Once again, please do break such PRs in multiple PRs or at least multiple commits.

Not by any means a full review, but I can see at least two outstanding issues with this PR:
- on my development environment, the advanced interface is broken with https://reactjs.org/docs/error-decoder.html?invariant=185 ; I'm not sure why, and I'm currently investigating
- profiles for remote users are routed client-side but not server-side: if you click on a remote user, say `@foo@bar`, the URL is `https://instance.com/@foo@bar` and refreshing will lead to a 404

Gargron committed:
Fix handling of remote usernames in routes

- When logged in, serve web app
- When logged out, redirect to permalink
- Fix `app-body` class not being set sometimes due to name conflict

Gargron committed:
Fix missing `multiColumn` prop

Gargron committed:
Fix failing test

Gargron committed:
Use `discoverable` attribute to control indexing directives

Gargron committed:
Fix `<ColumnLoading />` not using `multiColumn`

Gargron committed:
Add `noindex` to accounts in REST API

ClearlyClaire reviewed (commented):
I can confirm the crash has been fixed. The routing situation is better, but still leads to the awkward situation that you can navigate to `/@foo@bar` (a remote account displayed in the WebUI) or `/@foo@bar/1234` (a remote status displayed in the WebUI) from the logged-out WebUI but refreshing the page (or re-loading an unloaded page) will unexpectedly redirect you to the remote server.

I'm not sure what the best solution is here, as all the alternatives I can imagine have their own drawback:
- not redirecting allows sharing links to the cached version of a post
- opening the remote content in a new tab rather than allowing navigation to the cached version of it may not be as smooth as the current logged-out UI is. But maybe it's not so bad, that's what we had until you changed the logged-out UI to use the WebUI

Apart from that, I'm confused about some of the `noindex` uses. I have added inline comments.

I still have more general reservations about getting rid of server-side-rendered pages, in that:
- pages that did not require Javascript now require a good amount of it being loaded before showing anything. This means longer loading times, and it also means people disabling Javascript or with an incompatible browser (while we strive to maintain compatibility with a large amount of browsers, incompatibilities do occur) just won't see posts that people may share with them
- while I do think this reduces confusion about there being two views for you on your own server, I'm afraid it would bring *more* confusion when visiting remote servers. But maybe I'm wrong about that, and I'm not sure there's any way to know for sure except trying it at scale…

Gargron commented on app/javascript/mastodon/containers/mastodon.js:
```diff
@@ -78,13 +78,15 @@ export default class Mastodon extends React.PureComponent {
             </BrowserRouter>
 
-            <Helmet defaultTitle={title} titleTemplate={`%s - ${title}`} />
+            <Helmet defaultTitle={title} titleTemplate={`%s - ${title}`}>
+              <meta name='robots' content='noindex' />
+            </Helmet>
```
Hm, I thought of it as the default that would get overwritten by components that allow indexing. I guess it depends on if there's an intermediary state where this gets rendered on a route that would allow indexing otherwise, and whether Google's rendering engine waits for DOM to stop updating or not.

ClearlyClaire replied:
Right, though I believe the `Status` and `AccountTimeline` components are loaded asynchronously, so they probably will cause `noindex` to be added and committed to the DOM in most cases, which I believe would stop Google from indexing anything.

Gargron commented on app/controllers/about_controller.rb:
```diff
@@ -5,7 +5,15 @@ class AboutController < ApplicationController
 
   skip_before_action :require_functional!
 
+  before_action :set_instance_presenter
+
   def show
     expires_in 0, public: true unless user_signed_in?
   end
+
+  private
+
+  def set_instance_presenter
+    @instance_presenter = InstancePresenter.new
+  end
```
There was a crash because this presenter is used by the OpenGraph partial

ClearlyClaire replied:
ok, but I'm not sure how adding the OpenGraph partial to this controller is related to begin with

Gargron commented on app/javascript/mastodon/components/missing_indicator.js:
```diff
@@ -14,6 +15,10 @@ const MissingIndicator = ({ fullPage }) => (
       <FormattedMessage id='missing_indicator.label' tagName='strong' defaultMessage='Not found' />
       <FormattedMessage id='missing_indicator.sublabel' defaultMessage='This resource could not be found' />
     </div>
+
+    <Helmet>
+      <meta name='robots' content='noindex' />
+    </Helmet>
```
Good point, loading indicator for the status page would be a good thing...

Gargron commented on app/models/account.rb:
```diff
@@ -134,6 +134,7 @@ class Account < ApplicationRecord
            :role,
            :locale,
            :shows_application?,
+           :prefers_noindex?,
```
I used this wording because there's no actual way to enforce whether a search engine indexes something or not.

ClearlyClaire replied:
Right, but you can ensure you output `noindex`.

Gargron committed:
Change noindex directive to not be rendered by default before a route is mounted

Gargron committed:
Add loading indicator for detailed status in web UI

ClearlyClaire reviewed (commented):
My concerns about requiring Javascript, as well as refreshing the WebUI unexpectedly redirecting to remote URLs remain. The issues with `noindex` for statuses seem solved, but that's not the case for accounts.

Gargron committed:
Fix missing indicator appearing while account is loading in web UI

ClearlyClaire reviewed (commented):
I believe the issues with `noindex` have been resolved, my only remaining concerns are refreshing the WebUI unexpectedly redirecting to remote URLs, and more general concerns about requiring the use of Javascript.

busterb commented:
Howdy. Was trying to understand what the 'Feature on profile' option does in the UI, and stumbled across the endorsements API. After some backtracking through the tree, it seems that this change accidentally deleted the visualization from the UI:

https://github.com/mastodon/mastodon/commit/839f893168ab221b08fa439012189e6c29a2721a#r90342981

Seems this feature is no more as of this change, or it's not in a place I can find it :) : https://github.com/mastodon/mastodon/pull/8146

ClearlyClaire commented:
Yeah, it's currently not anywhere, unfortunately :/

ncfavier commented:
> * Remove remote follow / remote interaction dialogs (superceded by in-app interaction modal)

I don't understand this at all. The new interaction modal requires one to open a new tab/window to their own instance, copy the given URL, paste it into the search box before they can interact with any post. This is much, much less practical than the old modals that just asked me for my handle and then **remembered it**, so that any interaction takes two clicks.

Please at least provide a way to revert to the old behaviour, which should be the default IMO.
